src/libraries/System.Text.Json/src/System/Text/Json/AppContextSwitchHelper.cs

```diff
@@ -0,0 +1,45 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace System.Text.Json
+{
+    internal static class AppContextSwitchHelper
+    {
+        private const string SourceGenReflectionFallbackEnabled_SwitchName = "System.Text.Json.Serialization.EnableSourceGenReflectionFallback";
+        private static int s_SourceGenReflectionFallbackEnabled_CachedValue;
+
+        public static bool IsSourceGenReflectionFallbackEnabled
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get => GetCachedSwitchValue(SourceGenReflectionFallbackEnabled_SwitchName, ref s_SourceGenReflectionFallbackEnabled_CachedValue, defaultValue: false);
+        }
+
+        // Returns value of given switch using provided cache.
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool GetCachedSwitchValue(string switchName, ref int cachedSwitchValue, bool defaultValue)
+        {
+            // The cached switch value has 3 states: 0 - unknown, 1 - true, -1 - false
```

Comment:
I assume we do this elsewhere?

Reply:
Yes, this code has been lifted from the BinaryFormatter implementation.